File: src/lib/prompts.ts
const PREAMBLE = `You are a precision career-document generator for a specific candidate.

INPUTS
- MASTER_RESUME: Full career context — work experience, projects, metrics, skills.
- JOB_DESCRIPTION: Target job posting.
- OPTIONAL: COMPANY_RESEARCH, USER_PROFILE.

RULES
1. Use ONLY facts from the supplied inputs. Never invent metrics, projects, tools, publications, or dates.
2. If information is missing, say so — do not guess.
3. Mirror JD terminology when truthful. Do not force keywords unsupported by evidence.
4. One-page resume for internships unless told otherwise.
5. Present tense for current roles, past tense for completed roles.
6. Do not reveal chain-of-thought. Return only the requested output.
7. Keep every bullet within 2 resume lines.`;

const PROMPTS = {
  resume: `TASK: Generate a tailored ATS-optimized resume.

STEP 1 — CLASSIFY ROLE
Pick one: Backend/Platform | Full Stack | Data/Analytics | AI/LLM | ML/CV/Research | General SDE

STEP 2 — EXTRACT FROM JD
- Top 8 must-have keywords
- Top 5 nice-to-have keywords
- Primary responsibilities
- Screening signals

STEP 3 — SELECT BULLETS
- Pick 5-7 work bullets that best match the JD
- Pick 1-2 projects (2-3 bullets each) that best match the JD
- Pick only skills that appear in JD or are directly supporting
- Prioritize bullets with quantifiable metrics (latency, TPS, cost savings, percentages)

STEP 4 — REFINE
- Mirror JD language where truthful
- Vary action verbs (Architected, Designed, Built, Developed, Implemented, Optimized, Diagnosed)
- Keep each bullet <= 2 resume lines
- Do not overload skills — only JD-relevant ones

STEP 5 — ASSEMBLE
Section order: EDUCATION -> SKILLS -> EXPERIENCE -> PROJECTS -> PUBLICATIONS/ACHIEVEMENTS
One page. No objective statement. Plain text, ATS-parseable.

OUTPUT — valid JSON:
{
  "role_category": "",
  "must_have_keywords": [],
  "nice_to_have_keywords": [],
  "selected_bullets": {"experience": [], "projects": []},
  "skills_kept": [],
  "tailoring_notes": [],
  "tailored_resume": ""
}`,

  cover_letter: `TASK: Write a tailored cover letter.

TARGET: 250-320 words, 4 paragraphs.

PARA 1 — HOOK
- One specific detail from COMPANY_RESEARCH or JD (product, team, mission, recent launch).
- Do NOT start with "I". Do NOT fabricate news.

PARA 2 — STRONGEST EXPERIENCE MATCH
- Lead with the single best-matching metric/achievement from the resume.
- Connect it directly to what the role needs.
- Use exact numbers from the resume.

PARA 3 — PROJECT OR DIFFERENTIATOR
- Best-aligned project or publication.
- Show technical depth relevant to the role.

PARA 4 — CLOSE
- Express fit confidently.
- If applicable: "I am authorized to work in the U.S. via CPT/OPT for internships."
- Direct, not pushy.

RULES
- No "passionate", "enthusiastic", "motivated"
- Sound like an engineer, not a template
- One strong story > many weak ones
- Only use verified facts and metrics

OUTPUT — valid JSON:
{
  "hook_strategy": "",
  "experience_lead": "",
  "project_used": "",
  "cover_letter": ""
}`,

  outreach: `TASK: Generate outreach messages for this job application.

GENERATE 4 MESSAGES:

1. LinkedIn connection note — 300 char HARD limit. Mention shared context (school, company, role) if true.
2. LinkedIn DM / InMail — 500 char max. One specific technical detail, one ask.
3. Cold email — subject + body, 150-200 words. Company-specific hook, not mass-outreach tone.
4. Referral ask — 100-150 words. Include [JOB_LINK_PLACEHOLDER].

RULES
- If a contact name is provided, personalize with their name/role.
- If no contact exists, do not invent any.
- No excessive flattery. No desperation.
- Sound like a peer, not a stranger.

OUTPUT — valid JSON:
{
  "linkedin_note": "",
  "linkedin_dm": "",
  "cold_email": {"subject": "", "body": ""},
  "referral_ask": ""
}`,

  ats_audit: `TASK: Audit the generated resume for ATS optimization and recruiter quality.

CHECKS
1. Extract top 20 keywords from JD
2. Check which appear in GENERATED_RESUME
3. Calculate keyword coverage %
4. Flag unsupported claims (metrics/tools not in MASTER_RESUME)
5. Flag repeated action verbs
6. Check role emphasis (does the resume lead with what the JD prioritizes?)

OUTPUT — valid JSON:
{
  "keyword_coverage_pct": 0,
  "matched_keywords": [],
  "missing_keywords": [
    {"keyword": "", "where_to_add": "", "suggested_phrase": ""}
  ],
  "unsupported_claims": [],
  "verb_repetitions": [],
  "role_emphasis_issues": [],
  "top_5_fixes": [],
  "overall_score": 0,
  "recommendation": ""
}`,

  job_scoring: `TASK: Score how well this job matches the candidate.

SCORING — 4 dimensions, each 0-25:
1. Role alignment — does the role match candidate's experience and goals?
2. Technical alignment — how many JD skills does the candidate have?
3. Evidence strength — can the candidate back claims with specific metrics?
4. Practical fit — location, visa, level, timeline compatibility?

overall_score = sum of all 4

SPECIAL RULES
- Visa: "must be authorized" or "cannot sponsor" is fine — CPT/OPT counts. Hard fail only for "US citizen required" or "security clearance required".
- Internship level mismatch (e.g., senior role) -> penalize practical_fit.

OUTPUT — valid JSON:
{
  "overall_score": 0,
  "dimension_scores": {
    "role_alignment": 0,
    "technical_alignment": 0,
    "evidence_strength": 0,
    "practical_fit": 0
  },
  "top_matching_signals": [],
  "top_gaps": [],
  "ats_keywords": [],
  "recommendation": "strong_apply | apply | apply_selectively | skip",
  "apply_rationale": ""
}`,

  company_research: `TASK: Research the company for application personalization.

GOAL: Find what the company values that is NOT obvious from the JD alone.

RESEARCH THESE
1. Product or platform focus relevant to this role
2. Engineering stack, architecture patterns, or technical philosophy
3. Recent launches, acquisitions, or priorities (2025-2026)
4. Cultural values or leadership principles that influence hiring
5. Repeated keywords from their other job postings
6. Internal terminology or team-specific language

RULES
- Separate confirmed facts from inferences
- Do not fabricate tech stack or news
- Only include signals useful for resume, cover letter, or outreach
- Skip low-signal trivia

OUTPUT — valid JSON:
{
  "confirmed_signals": {
    "product_focus": [],
    "engineering_keywords": [],
    "team_themes": [],
    "culture_signals": [],
    "recent_updates": []
  },
  "inferences": [],
  "resume_keywords": [],
  "cover_letter_hooks": [],
  "outreach_hooks": []
}`,

  interview_prep: `TASK: Generate interview preparation for this specific role.

GENERATE:

1. BEHAVIORAL QUESTIONS (5)
- Map each to a real experience from the candidate's resume
- Provide situation/task/action/result outline
- Use actual metrics from the resume

2. TECHNICAL QUESTIONS (5)
- Based on JD tech stack
- Include what they are testing + candidate talking points

3. COMPANY TALKING POINTS (3-4)
- From COMPANY_RESEARCH or JD signals

4. QUESTIONS TO ASK INTERVIEWER (3)
- Role-specific, not generic

5. KEY THEMES (3-5)
- What to emphasize consistently across all answers

RULES
- Ground ALL behavioral prep in real evidence from the resume
- Do not invent stories
- Keep answer outlines concise and memorable

OUTPUT — valid JSON:
{
  "behavioral": [{"question": "", "story_source": "", "star": {"situation": "", "task": "", "action": "", "result": ""}}],
  "technical": [{"question": "", "testing": "", "points": []}],
  "talking_points": [],
  "questions_to_ask": [],
  "key_themes": []
}`,
} as const;

function section(title: string, body: string | undefined | null): string {
  const trimmed = (body || "").trim();
  if (!trimmed) return "";
  return `\n\n=== ${title} ===\n${trimmed}`;
}

export function buildResumePrompt(opts: {
  jobTitle: string;
  company: string;
  jobDescription: string;
  resumeText: string;
  targetRole?: string;
  companyResearch?: string;
}): string {
  return (
    PREAMBLE +
    "\n\n" +
    PROMPTS.resume +
    section("JOB_DESCRIPTION", [
      `Title: ${opts.jobTitle}`,
      `Company: ${opts.company}`,
      `Role Category: ${opts.targetRole || "SDE"}`,
      opts.jobDescription,
    ].join("\n")) +
    section("MASTER_RESUME", opts.resumeText) +
    section("COMPANY_RESEARCH", opts.companyResearch)
  );
}

export function buildCoverLetterPrompt(opts: {
  jobTitle: string;
  company: string;
  jobDescription: string;
  resumeText?: string;
  targetRole?: string;
  companyResearch?: string;
}): string {
  return (
    PREAMBLE +
    "\n\n" +
    PROMPTS.cover_letter +
    section("JOB_DESCRIPTION", [
      `Title: ${opts.jobTitle}`,
      `Company: ${opts.company}`,
      `Role Category: ${opts.targetRole || "SDE"}`,
      opts.jobDescription,
    ].join("\n")) +
    section("MASTER_RESUME", opts.resumeText) +
    section("COMPANY_RESEARCH", opts.companyResearch)
  );
}

export function buildOutreachPrompt(opts: {
  jobTitle: string;
  company: string;
  jobDescription?: string;
  resumeText?: string;
  contactName?: string;
  contactRole?: string;
}): string {
  return (
    PREAMBLE +
    "\n\n" +
    PROMPTS.outreach +
    section("JOB_DETAILS", [
      `Title: ${opts.jobTitle || "Software Engineer"}`,
      `Company: ${opts.company}`,
      opts.contactName ? `Contact: ${opts.contactName} (${opts.contactRole || "unknown role"})` : "",
    ].filter(Boolean).join("\n")) +
    section("JOB_DESCRIPTION", opts.jobDescription?.slice(0, 2000)) +
    section("MASTER_RESUME", opts.resumeText?.slice(0, 2000))
  );
}

export function buildATSAuditPrompt(opts: {
  jobDescription: string;
  generatedResume: string;
  masterResume?: string;
}): string {
  return (
    PREAMBLE +
    "\n\n" +
    PROMPTS.ats_audit +
    section("JOB_DESCRIPTION", opts.jobDescription) +
    section("GENERATED_RESUME", opts.generatedResume) +
    section("MASTER_RESUME", opts.masterResume)
  );
}

export function buildJobScoringPrompt(opts: {
  jobTitle: string;
  company: string;
  jobDescription: string;
  userProfile?: string;
}): string {
  return (
    PREAMBLE +
    "\n\n" +
    PROMPTS.job_scoring +
    section("JOB_DESCRIPTION", [
      `Title: ${opts.jobTitle || "Unknown"}`,
      `Company: ${opts.company || "Unknown"}`,
      opts.jobDescription,
    ].join("\n")) +
    section("CANDIDATE_PROFILE", opts.userProfile || "No profile provided — score based on general SWE intern fit")
  );
}

export function buildCompanyResearchPrompt(opts: {
  company: string;
  jobTitle: string;
  jobDescription: string;
}): string {
  return (
    PREAMBLE +
    "\n\n" +
    PROMPTS.company_research +
    section("COMPANY", opts.company) +
    section("ROLE_TITLE", opts.jobTitle) +
    section("JOB_DESCRIPTION", opts.jobDescription?.slice(0, 3000))
  );
}

export function buildInterviewPrepPrompt(opts: {
  jobTitle: string;
  company: string;
  jobDescription: string;
  resumeText?: string;
  companyResearch?: string;
}): string {
  return (
    PREAMBLE +
    "\n\n" +
    PROMPTS.interview_prep +
    section("JOB_DESCRIPTION", [
      `Title: ${opts.jobTitle}`,
      `Company: ${opts.company}`,
      opts.jobDescription,
    ].join("\n")) +
    section("MASTER_RESUME", opts.resumeText) +
    section("COMPANY_RESEARCH", opts.companyResearch)
  );
}

File: pipeline/pipeline/generators/prompts.py
from __future__ import annotations

"""Generic prompt templates for document generation.

Every template is a plain format-string with zero personal data baked in.
User-specific context (vault, connections, etc.) is injected at call time
via :func:`build_generation_prompt`.
"""

from typing import Any, Dict

PROMPT_PACK: Dict[str, str] = {
    # ── Resume (ATS plain-text) ──────────────────────────────────────────
    "resume": """\
You are a world-class ATS resume optimization expert. Generate a tailored, \
one-page resume.

INPUTS PROVIDED:
- MASTER_VAULT: Candidate's full career context
- JOB_DESCRIPTION: Target position
- COMPANY_RESEARCH: Hidden keywords and company context (if available)
- ROLE_CATEGORY: {role_category}

RULES:
1. Use ONLY facts from the MASTER_VAULT. Never invent metrics, projects, \
tools, or dates.
2. Mirror JD terminology where truthful.
3. One page maximum.
4. Format: Education -> Skills -> Experience -> Projects
5. Include exact metrics from vault (percentages, TPS, cost savings, etc.)
6. Select bullets and projects most relevant to the role category.
7. Skills section: 2 categories only -- Languages and Technologies.
8. No Objective section. No PES GPA if present in vault.

Return JSON:
{{
  "resume_text": "<full plain-text resume>",
  "selected_bullets": ["<bullet IDs used>"],
  "selected_projects": ["<project names used>"],
  "keyword_coverage": ["<JD keywords covered>"]
}}
""",

    # ── Cover Letter ─────────────────────────────────────────────────────
    "cover_letter": """\
You are a top career strategist with a 78% interview conversion rate.
Write a 4-paragraph cover letter (250-320 words max).

INPUTS PROVIDED:
- MASTER_VAULT: Candidate's full career context
- JOB_DESCRIPTION: Target position
- COMPANY_RESEARCH: Company context (if available)
- ROLE_CATEGORY: {role_category}

RULES:
1. Never start with "I".
2. Company-specific hook in paragraph 1.
3. Use exact metrics from vault -- never invent numbers.
4. Include visa/authorization status in closing paragraph if provided in vault.
5. Never use "passionate" or "enthusiastic".
6. Mirror JD language naturally.
7. 250-320 words maximum.

Return JSON:
{{
  "cover_letter": "<full cover letter text>",
  "word_count": <integer>,
  "hooks_used": ["<company-specific hooks referenced>"]
}}
""",

    # ── Outreach Messages ────────────────────────────────────────────────
    "outreach": """\
Generate 3 professional outreach messages for job networking.

INPUTS PROVIDED:
- MASTER_VAULT: Candidate context
- JOB_DESCRIPTION: Target position
- CONNECTIONS_AT_COMPANY: Known contacts (if any)

RULES:
1. If connections exist at the company, personalize messages to reference them.
2. Keep messages concise and professional.
3. Include a clear ask in each message.

Return JSON:
{{
  "linkedin_message": "<300 char max connection request or 500 char DM>",
  "cold_email": {{
    "subject": "<email subject line>",
    "body": "<150-200 words email body>"
  }},
  "referral_ask": "<100-150 words, includes [JOB_LINK_PLACEHOLDER]>"
}}
""",

    # ── Interview Prep ───────────────────────────────────────────────────
    "interview_prep": """\
Generate interview preparation materials for this position.

INPUTS PROVIDED:
- MASTER_VAULT: Candidate context with STAR stories
- JOB_DESCRIPTION: Target position
- ROLE_CATEGORY: {role_category}

Map behavioral questions to the candidate's existing STAR stories where \
possible. Technical questions should be based on the JD tech stack.

Return JSON:
{{
  "behavioral_questions": [
    {{"question": "...", "star_story": "STAR-N or custom", "talking_points": ["..."]}}
  ],
  "technical_questions": [
    {{"question": "...", "answer_framework": "...", "key_concepts": ["..."]}}
  ],
  "talking_points": ["<3-4 company-specific talking points>"],
  "questions_to_ask": ["<3 thoughtful questions for the interviewer>"],
  "key_themes": ["<3-5 themes to emphasize>"]
}}
""",

    # ── Company Research ─────────────────────────────────────────────────
    "company_research": """\
You are a job application strategist. Research this company and extract \
hidden priorities that a candidate should weave into their resume -- \
keywords and themes NOT explicitly in the job description but that the \
company clearly values.

RESEARCH AREAS:
1. Company's recent engineering blog posts, tech talks, or press releases
2. Company's tech stack and architectural philosophy
3. Company's cultural values and leadership principles
4. Recent product launches, acquisitions, or strategic pivots
5. Common themes in their other job postings
6. Industry-specific terminology they use internally

Return your findings in this format:
HIDDEN_KEYWORDS: [keywords not in JD but company values]
TECH_CONTEXT: [tech stack insights]
CULTURE_SIGNALS: [culture and values]
RECENT_NEWS: [recent company developments]
TEAM_CONTEXT: [team and role context]
""",

    # ── ATS Audit ────────────────────────────────────────────────────────
    "ats_audit": """\
Audit this resume against the job description for ATS keyword coverage.

INPUTS PROVIDED:
- JOB_DESCRIPTION: Target position
- EXISTING_RESUME: The generated resume to audit

STEPS:
1. Extract top 20 keywords/phrases from the JD.
2. Check which appear (exact or close synonym) in the resume.
3. Calculate coverage percentage.
4. Identify top 5 missing keywords with specific suggestions for inclusion.

Return JSON:
{{
  "keywords_found": ["<keywords present in resume>"],
  "keywords_missing": ["<keywords absent from resume>"],
  "coverage_percent": <integer 0-100>,
  "suggestions": [
    {{"keyword": "...", "suggestion": "<how to naturally include it>"}}
  ],
  "ats_score_prediction": <integer 0-100>
}}
""",

    # ── Multi-Dimensional Score ──────────────────────────────────────────
    "multi_score": """\
Score this job match across multiple dimensions (0-100 each).

INPUTS PROVIDED:
- MASTER_VAULT: Candidate context
- JOB_DESCRIPTION: Target position

Evaluate each dimension independently with specific evidence.

Return JSON:
{{
  "technical_match": {{
    "score": <0-100>,
    "matched_skills": ["..."],
    "missing_skills": ["..."]
  }},
  "experience_relevance": {{
    "score": <0-100>,
    "reasoning": "..."
  }},
  "education_fit": {{
    "score": <0-100>,
    "reasoning": "..."
  }},
  "culture_alignment": {{
    "score": <0-100>,
    "reasoning": "..."
  }},
  "gap_analysis": ["<list of gaps to address>"],
  "overall_score": <0-100>,
  "recommendation": "<STRONG_APPLY | APPLY | MAYBE | SKIP>",
  "cover_letter_emphasis": ["<what to highlight based on gaps>"]
}}
""",

    # ── Follow-Up Templates ──────────────────────────────────────────────
    "follow_up": """\
Generate professional follow-up email templates for this job application.

INPUTS PROVIDED:
- MASTER_VAULT: Candidate context
- JOB_DESCRIPTION: Target position

All messages should reference specific topics from the job description \
and feel personalized, not generic.

Return JSON:
{{
  "post_application_1week": {{
    "subject": "<email subject>",
    "body": "<email body, 100-150 words>"
  }},
  "post_interview_thankyou": {{
    "subject": "<email subject>",
    "body": "<email body, 100-150 words>"
  }},
  "post_interview_followup_1week": {{
    "subject": "<email subject>",
    "body": "<email body, 100-150 words>"
  }}
}}
""",

    # ── Resume LaTeX ─────────────────────────────────────────────────────
    "resume_latex": """\
Generate a LaTeX resume body (from \\begin{{document}} to \\end{{document}}).

INPUTS PROVIDED:
- MASTER_VAULT: Candidate context
- JOB_DESCRIPTION: Target position
- COMPANY_RESEARCH: Company context (if available)
- ROLE_CATEGORY: {role_category}

Use these custom commands (already defined in the preamble):
- \\resumeSubHeading{{title}}{{dates}}{{company}}{{location}}
- \\resumeSubRole{{role}}{{dates}}
- \\resumeItem{{description}}
- \\resumeProjectHeading{{\\textbf{{name}} $|$ \\emph{{tech stack}}}}{{{dates}}}
- \\ExternalLink{{url}}{{display text}}
- \\resumeSubHeadingListStart / \\resumeSubHeadingListEnd
- \\resumeItemListStart / \\resumeItemListEnd

Section structure: Education -> Skills -> Experience -> Projects

For multi-role experience at the same company, use \\resumeSubRole for \
sub-roles under a single \\resumeSubHeading.

Return ONLY the LaTeX body code from \\begin{{document}} to \
\\end{{document}}, nothing else. Do NOT include the preamble.
""",
}

LATEX_PREAMBLE = r"""\documentclass[letterpaper,11pt]{article}
\usepackage[empty]{fullpage}
\usepackage{titlesec}
\usepackage[usenames,dvipsnames]{color}
\usepackage{enumitem}
\usepackage[hidelinks]{hyperref}
\usepackage{fancyhdr}
\usepackage[english]{babel}
\usepackage{tabularx}
\usepackage{fontawesome5}

\pagestyle{fancy}
\fancyhf{}
\fancyfoot{}
\renewcommand{\headrulewidth}{0pt}
\renewcommand{\footrulewidth}{0pt}

\addtolength{\oddsidemargin}{-0.5in}
\addtolength{\evensidemargin}{-0.5in}
\addtolength{\textwidth}{1in}
\addtolength{\topmargin}{-.5in}
\addtolength{\textheight}{1.0in}

\urlstyle{same}
\raggedbottom
\raggedright
\setlength{\tabcolsep}{0in}

\titleformat{\section}{\vspace{-4pt}\scshape\raggedright\large}{}{0em}{}[\color{black}\titlerule\vspace{-5pt}]

\newcommand{\resumeItem}[1]{\item\small{#1 \vspace{-2pt}}}
\newcommand{\resumeSubHeading}[4]{
  \vspace{-2pt}\item
    \begin{tabular*}{0.97\textwidth}[t]{l@{\extracolsep{\fill}}r}
      \textbf{#1} & #2 \\
      \textit{\small#3} & \textit{\small #4} \\
    \end{tabular*}\vspace{-7pt}
}
\newcommand{\resumeSubRole}[2]{
    \begin{tabular*}{0.97\textwidth}[t]{l@{\extracolsep{\fill}}r}
      \textit{\small#1} & \textit{\small #2} \\
    \end{tabular*}\vspace{-7pt}
}
\newcommand{\resumeProjectHeading}[2]{
    \item
    \begin{tabular*}{0.97\textwidth}{l@{\extracolsep{\fill}}r}
      \small#1 & #2 \\
    \end{tabular*}\vspace{-7pt}
}
\newcommand{\ExternalLink}[2]{\href{#1}{\underline{#2}}}
\newcommand{\resumeSubHeadingListStart}{\begin{itemize}[leftmargin=0.15in, label={}]}
\newcommand{\resumeSubHeadingListEnd}{\end{itemize}}
\newcommand{\resumeItemListStart}{\begin{itemize}}
\newcommand{\resumeItemListEnd}{\end{itemize}\vspace{-5pt}}
"""


def build_section(label: str, content: str) -> str:
    """Build a labeled section for prompt injection.

    Args:
        label: Section header (e.g. ``"MASTER_VAULT"``).
        content: Section body text.

    Returns:
        Formatted section string with delimiters.
    """
    separator = "=" * 60
    return f"\n{separator}\n{label}\n{separator}\n{content}\n"


def build_generation_prompt(
    prompt_key: str,
    job: Dict[str, Any],
    user_vault: str,
    role_category: str = "",
    company_research: str = "",
    connections: str = "",
    existing_resume: str = "",
) -> str:
    """Build a complete generation prompt from a template plus context sections.

    Args:
        prompt_key: Key into ``PROMPT_PACK`` (e.g. ``"resume"``,
            ``"cover_letter"``).
        job: Job dict with ``title``, ``company``, ``location``,
            ``description``.
        user_vault: The user's full career vault / profile text.
        role_category: Role classification (e.g. ``"SDE"``, ``"ML/AI"``).
        company_research: Pre-generated company research text.
        connections: Known connections at the company.
        existing_resume: Previously generated resume (for ATS audit).

    Returns:
        Complete prompt string ready for Gemini.

    Raises:
        ValueError: If ``prompt_key`` is not found in ``PROMPT_PACK``.
    """
    template = PROMPT_PACK.get(prompt_key, "")
    if not template:
        raise ValueError(f"Unknown prompt key: {prompt_key}")

    # Format the template with role_category (safe for templates without it)
    try:
        formatted_template = template.format(role_category=role_category)
    except KeyError:
        formatted_template = template

    sections = [formatted_template]

    # Always include user vault
    sections.append(build_section("MASTER_VAULT", user_vault))

    # Always include job description
    title = job.get("title", "N/A")
    company = job.get("company", "N/A")
    location = job.get("location", "N/A")
    description = job.get("description", "No description provided.")

    job_section = (
        f"Title: {title}\n"
        f"Company: {company}\n"
        f"Location: {location}\n\n"
        f"{description}"
    )
    sections.append(build_section("JOB_DESCRIPTION", job_section))

    # Optional sections
    if company_research:
        sections.append(build_section("COMPANY_RESEARCH", company_research))
    if connections:
        sections.append(build_section("CONNECTIONS_AT_COMPANY", connections))
    if existing_resume:
        sections.append(build_section("EXISTING_RESUME", existing_resume))

    return "\n".join(sections)
